allennlp/training/metrics/conll_coref_scores.py

```diff
@@ -0,0 +1,155 @@
+                                                                                      predicted_antecedents)
+        for i, metadata in enumerate(metadata_list):
+            gold_clusters, mention_to_gold = self.get_gold_clusters(metadata["clusters"])
+            predicted_clusters, mention_to_predicted = self.get_predicted_clusters(top_spans[i],
+                                                                                   antecedent_indices,
+                                                                                   predicted_antecedents[i])
+            for scorer in self.scorers:
+                scorer.update(predicted_clusters, gold_clusters, mention_to_predicted, mention_to_gold)
+
+    @overrides
+    def get_metric(self, reset: bool = False) -> Tuple[float, float, float]:
+        metrics = (lambda e: e.get_precision(), lambda e: e.get_recall(), lambda e: e.get_f1())
+        precision, recall, f1_score = tuple(sum(m(e) for e in self.scorers) / len(self.scorers) for m in metrics)
+        if reset:
+            self.reset()
+        return precision, recall, f1_score
+
+    @overrides
+    def reset(self):
+        self.scorers = [Scorer(m) for m in (Scorer.muc, Scorer.b_cubed, Scorer.ceafe)]
+
+    @staticmethod
+    def get_gold_clusters(gold_clusters):
+        gold_clusters = [tuple(tuple(m) for m in gc) for gc in gold_clusters]
+        mention_to_gold = {}
+        for gold_cluster in gold_clusters:
+            for mention in gold_cluster:
+                mention_to_gold[mention] = gold_cluster
+        return gold_clusters, mention_to_gold
+
+    @staticmethod
+    def get_predicted_clusters(top_spans, antecedent_indices, predicted_antecedents):
+        mention_to_predicted = {}
+        predicted_clusters = []
+        for i, predicted_antecedent in enumerate(predicted_antecedents):
+            if predicted_antecedent < 0:
+                continue
+            predicted_index = antecedent_indices[i, predicted_antecedent]
+            assert i > predicted_index
+            predicted_antecedent = tuple(top_spans[predicted_index])
+            if predicted_antecedent in mention_to_predicted:
+                predicted_cluster = mention_to_predicted[predicted_antecedent]
+            else:
+                predicted_cluster = len(predicted_clusters)
+                predicted_clusters.append([predicted_antecedent])
+                mention_to_predicted[predicted_antecedent] = predicted_cluster
+            mention = tuple(top_spans[i])
+            predicted_clusters[predicted_cluster].append(mention)
+            mention_to_predicted[mention] = predicted_cluster
+        predicted_clusters = [tuple(pc) for pc in predicted_clusters]
+        mention_to_predicted = {m:predicted_clusters[i] for m, i in mention_to_predicted.items()}
+        return predicted_clusters, mention_to_predicted
+
+
+class Scorer(object):
```

Comment:
I know you just borrowed this class, but it would be great if you could clean up the variable naming etc. For some stuff I can guess what it means, but I actually don't know what `ceafe`, `muc` etc mean.

Reply:
ceafe and muc are names of coref metrics.  I don't think we need to change those names, but referencing them in the class docstring is probably warranted.

Reply:
Added reference of the various metrics to docstrings.